kikuchipy/indexation/similarity_metrics.py

```diff
@@ -0,0 +1,394 @@
+    """
+    metric_cls = FlatSimilarityMetric if flat else SimilarityMetric
+    p_scope_ndim, t_scope_ndim = metric_cls._SCOPE_TO_P_T_NDMI[
+        MetricScope.MANY_TO_MANY
+    ]
+    p = p[(np.newaxis,) * (p_scope_ndim - p.ndim)]
+    t = t[(np.newaxis,) * (t_scope_ndim - t.ndim)]
+    return p, t
+
+
+def zero_mean(p, t, flat=False):
+    """Subtract the mean from patterns and templates of any scope.
+
+    Parameters
+    ----------
+    p : Union[np.ndarray, da.Array]
+        Patterns
+    t : Union[np.ndarray, da.Array]
+        Templates
+    flat : bool, optional
+        Wheter `p` and `t` are flattened, by default False
+
+    Returns
+    -------
+    (p,t)
+        Tuple of p and t with their mean subtracted.
+    """
+    squeeze = 1 not in p.shape + t.shape
+    p, t = expand_dims_to_many_to_many(p, t, flat)
+    p_mean_axis = 1 if flat else (2, 3)
+    t_mean_axis = 1 if flat else (1, 2)
+    p -= p.mean(axis=p_mean_axis, keepdims=True)
+    t -= t.mean(axis=t_mean_axis, keepdims=True)
+
+    if squeeze:
+        return (
+            p.squeeze(),
+            t.squeeze(),
+        )
+    else:
+        return p, t
+
+
+def normalize(p, t, flat=False):
+    """Normalize patterns and templates of any scope.
+
+    Parameters
+    ----------
+    p : Union[np.ndarray, da.Array]
+        Patterns
+    t : Union[np.ndarray, da.Array]
+        Templates
+    flat : bool, optional
+        Wheter `p` and `t` are flattened, by default False
+
+    Returns
+    -------
+    (p,t)
+        Tuple of p and t divided respectively by their L2-norms.
+    """
+    squeeze = 1 not in p.shape + t.shape
+    p, t = expand_dims_to_many_to_many(p, t, flat)
+    p_sum_axis = 1 if flat else (2, 3)
+    t_sum_axis = 1 if flat else (1, 2)
+    p /= (p ** 2).sum(axis=p_sum_axis, keepdims=True) ** 0.5
+    t /= (t ** 2).sum(axis=t_sum_axis, keepdims=True) ** 0.5
+
+    if squeeze:
+        return (
+            p.squeeze(),
+            t.squeeze(),
+        )
+    else:
+        return (p, t)
+
+
+SIMILARITY_METRICS: Dict[str, Callable] = {}
+
+
+def _zncc_einsum(
+    patterns: Union[da.Array, np.ndarray],
+    templates: Union[da.Array, np.ndarray],
+):
+    """Compute (lazily) ZNCC between patterns and templates.
+
+    Parameters
+    ----------
+    patterns : Union[da.Array, np.ndarray]
+        [description]
+    templates : Union[da.Array, np.ndarray]
+        [description]
+
+    Returns
+    -------
+    Union[da.Array, np.ndarray]
+        [description]
+    """
+    patterns, templates = zero_mean(patterns, templates)
+    patterns, templates = normalize(patterns, templates)
+
+    zncc = da.einsum(
+        "ijk,lmjk->ilm", templates, patterns, optimize=True
+    )  # TODO: Will fail if np.ndarray!!!
+
+    # Alternative with equivalent results:
+    #   zncc = da.tensordot(templates, patterns, axes=([1, 2], [2, 3]))
+    return zncc
+
+
+SIMILARITY_METRICS["zncc"] = make_similarity_metric(
```

Comment:
Is the idea that the user will do this

```python
>>> from kikuchipy.indexing.similarity_metrics import SIMILARITY_METRIC
>>> zncc = SIMILARITY_METRIC["zncc"]
>>> ndp = SIMILARITY_METRIC["ndp"]
```

I find it cleaner with

```python
>>> from kikuchipy.indexing.similarity_metrics import zncc, ndp
```

What do you think?